File: spike/featurize.py
"""Sparse feature rows for the gpu-query tagger.

Two families of rows compete here, and the spike exists to find out which one
the model uses.

Mechanical rows carry word identity through hashes. They let the model memorize
that "status" names a field, and they are worthless on a schema it has never
seen.

Schema rows carry no identity at all. They say only that a token matched some
field of some kind, or matched an enum value belonging to the nearest preceding
field. A model leaning on these transfers to any schema.

`use_word_identity=False` folds the hash blocks onto the padding row, which is
the ablation that separates the two stories.
"""

from __future__ import annotations

from dataclasses import dataclass

import match
from schema import BOOL, DATE, ENUM, NUMBER, TEXT, Schema

SLOTS = 16

# Closed English operator lexicon. Not app-specific, so this is not a leak.
OPERATOR_WORDS = [
    "and",
    "or",
    "not",
    "no",
    "without",
    "excluding",
    "is",
    "before",
    "after",
    "since",
    "over",
    "under",
    "above",
    "below",
    "than",
    "least",
    "most",
    "more",
    "less",
    "fewer",
    "prior",
    "later",
    "earlier",
    "containing",
    "mentioning",
    "matching",
    "with",
    ":",
    "=",
]
OPERATOR_INDEX = {word: index for index, word in enumerate(OPERATOR_WORDS)}

KIND_ROWS = 3
LENGTH_ROWS = 8
CHAR_ROWS = 40
WORD_HASH_ROWS = 256
DEVOWEL_ROWS = 128
FLAG_ROWS = 4
FIELD_MATCH_ROWS = 6  # none, bool, enum, date, number, text
MATCH_QUALITY_ROWS = 5  # none, exact, stem, prefix, typo
ENUM_ANY_ROWS = 1
ENUM_PRECEDING_ROWS = 1
DISTANCE_ROWS = 6  # 0..4 tokens since a field, plus a distinct "no field yet"
OPERATOR_ROWS = len(OPERATOR_WORDS) + 1
POSITION_ROWS = 4

_blocks = [
    ("kind", KIND_ROWS),
    ("length", LENGTH_ROWS),
    ("first_char", CHAR_ROWS),
    ("last_char", CHAR_ROWS),
    ("word_hash", WORD_HASH_ROWS),
    ("devowel_hash", DEVOWEL_ROWS),
    ("flags", FLAG_ROWS),
    ("field_match", FIELD_MATCH_ROWS),
    ("match_quality", MATCH_QUALITY_ROWS),
    ("enum_any", ENUM_ANY_ROWS),
    ("enum_preceding", ENUM_PRECEDING_ROWS),
    ("distance", DISTANCE_ROWS),
    ("operator", OPERATOR_ROWS),
    ("position", POSITION_ROWS),
]

OFFSET: dict[str, int] = {}
_cursor = 0
for _name, _size in _blocks:
    OFFSET[_name] = _cursor
    _cursor += _size
FEATURE_ROWS = _cursor
PADDING_ROW = FEATURE_ROWS

IDENTITY_BLOCKS = ("word_hash", "devowel_hash")
KIND_OF_FIELD = {BOOL: 1, ENUM: 2, DATE: 3, NUMBER: 4, TEXT: 5}
LENGTH_BUCKETS = [1, 2, 3, 4, 6, 8, 12, 10**9]


def _hash(text: str) -> int:
    value = 2166136261
    for character in text:
        value = ((value ^ ord(character)) * 16777619) & 0xFFFFFFFF
    return value


def _character_class(character: str) -> int:
    code = ord(character.lower())
    if 97 <= code <= 122:
        return code - 97
    if 48 <= code <= 57:
        return code - 48 + 26
    return 36 + (code % 4)


def _length_bucket(text: str) -> int:
    for index, bound in enumerate(LENGTH_BUCKETS):
        if len(text) <= bound:
            return index
    return len(LENGTH_BUCKETS) - 1


@dataclass
class SchemaIndex:
    """Surface-form lookups for one schema, built once per example."""

    field_kind: dict[str, str]
    enum_owner: dict[str, set[str]]

    @classmethod
    def build(cls, active: Schema) -> "SchemaIndex":
        field_kind: dict[str, str] = {}
        enum_owner: dict[str, set[str]] = {}
        for item in active.fields:
            for surface in item.surface_forms():
                field_kind[surface.lower()] = item.kind
            for value in item.values:
                enum_owner.setdefault(value.lower(), set()).add(item.name)
        return cls(field_kind=field_kind, enum_owner=enum_owner)


def rows_for(
    tokens: list[str], active: Schema, use_word_identity: bool = True
) -> list[list[int]]:
    """One padded row list per token."""
    index = SchemaIndex.build(active)
    entries = match.build_index(active)
    lowered = [token.lower() for token in tokens]
    resolved = [match.resolve(token, active, entries) for token in lowered]

    # Which field, if any, most recently appeared before each token. Canonical
    # names, so "assngnee" anchors the values after it exactly as "assignee" would.
    preceding_field: list[str | None] = []
    distance_to_field: list[int] = []
    current: str | None = None
    distance = 99
    for position in range(len(lowered)):
        preceding_field.append(current)
        distance_to_field.append(distance)
        if resolved[position] is not None:
            current = resolved[position].field
            distance = 0
        elif distance < 99:
            distance += 1

    output: list[list[int]] = []
    for position, token in enumerate(tokens):
        folded = lowered[position]
        rows: list[int] = []

        def add(block: str, value: int) -> None:
            if block in IDENTITY_BLOCKS and not use_word_identity:
                return
            rows.append(OFFSET[block] + value)

        if folded.replace(".", "").isdigit():
            kind = 1
        elif folded[0].isalpha() or folded[0] == "_":
            kind = 0
        else:
            kind = 2
        add("kind", kind)
        add("length", _length_bucket(token))
        add("first_char", _character_class(token[0]))
        add("last_char", _character_class(token[-1]))
        add("word_hash", _hash(folded) & 255)
        stripped = "".join(c for c in folded if c not in "aeiou") or folded
        add("devowel_hash", _hash(stripped) & 127)

        if token[:1].isupper():
            add("flags", 0)
        if any(c.isdigit() for c in token):
            add("flags", 1)
        if "-" in token or "_" in token:
            add("flags", 2)

        found = resolved[position]
        add("field_match", KIND_OF_FIELD[found.kind] if found else 0)
        add("match_quality", found.quality + 1 if found else 0)

        owners = index.enum_owner.get(folded)
        if owners:
            add("enum_any", 0)
            # Does this enum value belong to the field just named? This row is
            # what disambiguates a value two fields share.
            if preceding_field[position] in owners:
                add("enum_preceding", 0)

        gap = distance_to_field[position]
        add("distance", DISTANCE_ROWS - 1 if gap == 99 else min(gap, DISTANCE_ROWS - 2))
        add("operator", OPERATOR_INDEX.get(folded, len(OPERATOR_WORDS)))

        span = max(1, len(tokens) - 1)
        add("position", min(3, int(4 * position / (span + 1))))

        rows = rows[:SLOTS] + [PADDING_ROW] * max(0, SLOTS - len(rows))
        output.append(rows)
    return output


if __name__ == "__main__":
    import random

    import generate
    import schema as schema_module

    schema_module.assert_disjoint()
    print(f"FEATURE_ROWS = {FEATURE_ROWS}  (padding row {PADDING_ROW})")
    for name, size in _blocks:
        print(f"  {name:<16} {OFFSET[name]:>4} .. {OFFSET[name] + size - 1:>4}")
    estimated = FEATURE_ROWS * 32 + 14393
    print(f"\nparameters at HIDDEN=32: {FEATURE_ROWS} * 32 + 14393 = {estimated:,}")

    example = generate.dataset("eval", 1, seed=3)[0]
    print(f"\nexample: {example.text}")
    table = rows_for(example.tokens, example.schema)
    for token, label, row in zip(example.tokens, example.labels, table):
        active = [value for value in row if value != PADDING_ROW]
        named = []
        for value in active:
            for name, size in _blocks:
                if OFFSET[name] <= value < OFFSET[name] + size:
                    named.append(f"{name}+{value - OFFSET[name]}")
                    break
        print(f"  {token:<14} {generate.LABELS[label]:<12} {' '.join(named)}")

File: site/src/runtime/featurize.ts
/**
 * Sparse feature rows and neighbour indices.
 *
 * Ported from spike/featurize.py and `neighbors_for` in spike/train.py. Row
 * offsets must match the Python exactly, including the two word-identity blocks
 * that this checkpoint never emits: the blocks still occupy their offsets, they
 * are simply skipped. The shipped model was trained with --no-word-identity, so
 * nothing here carries a field name into the weights.
 */

import { buildIndex, enumOwners, resolve, surfaceForms } from "./schema";
import type { Index, Kind, Match, Schema } from "./schema";

export const SLOTS = 16;

const OPERATOR_WORDS = [
  "and", "or", "not", "no", "without", "excluding", "is", "before", "after",
  "since", "over", "under", "above", "below", "than", "least", "most", "more",
  "less", "fewer", "prior", "later", "earlier", "containing", "mentioning",
  "matching", "with", ":", "=",
];
const OPERATOR_INDEX = new Map(OPERATOR_WORDS.map((w, i) => [w, i]));

const BLOCKS: Array<[string, number]> = [
  ["kind", 3],
  ["length", 8],
  ["first_char", 40],
  ["last_char", 40],
  ["word_hash", 256],
  ["devowel_hash", 128],
  ["flags", 4],
  ["field_match", 6],
  ["match_quality", 5],
  ["enum_any", 1],
  ["enum_preceding", 1],
  ["distance", 6],
  ["operator", 30],
  ["position", 4],
];

export const OFFSET: Record<string, number> = {};
let cursor = 0;
for (const [name, size] of BLOCKS) {
  OFFSET[name] = cursor;
  cursor += size;
}
export const FEATURE_ROWS = cursor;
export const PADDING_ROW = FEATURE_ROWS;

const KIND_OF_FIELD: Record<Kind, number> = {
  bool: 1, enum: 2, date: 3, number: 4, text: 5,
};
const LENGTH_BUCKETS = [1, 2, 3, 4, 6, 8, 12, 1e9];
const DISTANCE_ROWS = 6;

function characterClass(character: string): number {
  const code = character.toLowerCase().charCodeAt(0);
  if (code >= 97 && code <= 122) return code - 97;
  if (code >= 48 && code <= 57) return code - 48 + 26;
  return 36 + (code % 4);
}

function lengthBucket(text: string): number {
  for (let i = 0; i < LENGTH_BUCKETS.length; i++) {
    if (text.length <= LENGTH_BUCKETS[i]) return i;
  }
  return LENGTH_BUCKETS.length - 1;
}

function isDigits(text: string): boolean {
  const stripped = text.replace(/\./g, "");
  return stripped.length > 0 && /^[0-9]+$/.test(stripped);
}

export interface Featurized {
  rows: number[][];
  neighbors: Array<[number, number]>;
  resolved: Array<Match | null>;
}

export function featurize(tokens: string[], schema: Schema): Featurized {
  const index: Index = buildIndex(schema);
  const owners = enumOwners(schema);
  const lowered = tokens.map((t) => t.toLowerCase());
  const resolved = lowered.map((t) => resolve(t, index));

  // Nearest resolved field before each token, by canonical name.
  const precedingField: Array<string | null> = [];
  const distanceToField: number[] = [];
  let current: string | null = null;
  let distance = 99;
  for (let i = 0; i < lowered.length; i++) {
    precedingField.push(current);
    distanceToField.push(distance);
    if (resolved[i]) {
      current = resolved[i]!.field;
      distance = 0;
    } else if (distance < 99) {
      distance += 1;
    }
  }

  const rows: number[][] = [];
  for (let position = 0; position < tokens.length; position++) {
    const token = tokens[position];
    const folded = lowered[position];
    const active: number[] = [];
    const add = (block: string, value: number) => {
      // word_hash and devowel_hash are the ablated blocks; never emitted.
      if (block === "word_hash" || block === "devowel_hash") return;
      active.push(OFFSET[block] + value);
    };

    let kind: number;
    if (isDigits(folded)) kind = 1;
    else if (/[A-Za-z_]/.test(folded[0])) kind = 0;
    else kind = 2;
    add("kind", kind);
    add("length", lengthBucket(token));
    add("first_char", characterClass(token[0]));
    add("last_char", characterClass(token[token.length - 1]));

    if (token[0] === token[0].toUpperCase() && /[A-Za-z]/.test(token[0])) add("flags", 0);
    if (/[0-9]/.test(token)) add("flags", 1);
    if (token.includes("-") || token.includes("_")) add("flags", 2);

    const found = resolved[position];
    add("field_match", found ? KIND_OF_FIELD[found.kind] : 0);
    add("match_quality", found ? found.quality + 1 : 0);

    const owning = owners.get(folded);
    if (owning) {
      add("enum_any", 0);
      const previous = precedingField[position];
      if (previous !== null && owning.has(previous)) add("enum_preceding", 0);
    }

    const gap = distanceToField[position];
    add("distance", gap === 99 ? DISTANCE_ROWS - 1 : Math.min(gap, DISTANCE_ROWS - 2));
    add("operator", OPERATOR_INDEX.get(folded) ?? OPERATOR_WORDS.length);

    const span = Math.max(1, tokens.length - 1);
    add("position", Math.min(3, Math.floor((4 * position) / (span + 1))));

    const padded = active.slice(0, SLOTS);
    while (padded.length < SLOTS) padded.push(PADDING_ROW);
    rows.push(padded);
  }

  return { rows, neighbors: neighborsFor(tokens, schema), resolved };
}

/**
 * Nearest field-matching token on each side.
 *
 * Note this uses EXACT surface matching, not the tolerant resolver, because the
 * Python it mirrors builds its index from surface forms directly.
 */
export function neighborsFor(tokens: string[], schema: Schema): Array<[number, number]> {
  const surfaces = new Set<string>();
  for (const field of schema) {
    for (const surface of surfaceForms(field)) surfaces.add(surface.toLowerCase());
  }
  const isField = tokens.map((t) => surfaces.has(t.toLowerCase()));

  const previous: number[] = new Array(tokens.length).fill(-1);
  let last = -1;
  for (let i = 0; i < tokens.length; i++) {
    previous[i] = last;
    if (isField[i]) last = i;
  }

  const following: number[] = new Array(tokens.length).fill(-1);
  let upcoming = -1;
  for (let i = tokens.length - 1; i >= 0; i--) {
    following[i] = upcoming;
    if (isField[i]) upcoming = i;
  }

  return tokens.map((_, i) => [previous[i], following[i]] as [number, number]);
}
